### DISCUSSION TITLE
[Feature request] 32-bit Docker images

### DISCUSSION THREAD

**Discussion by binaryfire** [valkey-io/valkey#1304] [General]
Hi all

I did some testing after reading the following in the docs and found a _significant_ reduction in memory usage with 32-bit builds:

> **Using 32-bit instances**
> When Valkey is compiled as a 32-bit target, it uses a lot less memory per key, since pointers are small, but such an instance will be limited to 4 GB of maximum memory usage. To compile Valkey as 32-bit binary use make 32bit. RDB and AOF files are compatible between 32-bit and 64-bit instances (and between little and big endian of course) so you can switch from 32 to 64-bit, or the contrary, without problems.
> 

Is there any chance official 32-bit versions could be added to Docker Hub? I realise I can build them myself but official images would make maintenance a lot easier. The use case is caching - the majority of web apps don't need anything close to 4GB and the reduced memory footprint is a big help in resource constrained environments.

**Comment by stockholmux** [comment]
I have mixed feelings about this, so I did some tests to find out the difference between 32-Bit Valkey vs 64-Bit Valkey.

Here is what I found:
-  Building 32-bit Valkey on a modern distro is not easy. I tried it on Fedora 41 and I had to install a lot of random packages compared to 64-bit which just kinda works out-of-the-box.
- The difference is in memory efficiency big, but not huge. Here was test (feel free to tell me I did it wrong).
   - I filled each instance by running `DEBUG POPULATE 2500 test 1000000` (2500 keys at ~1mb each)
   - I ran `sudo pmap -x <pid>`
   - Ran this on x86_64 Fedora 41 each on 32-bit Valkey and as a 64-bit Valkey with a standard valkey.conf, only modified with `enable-debug-command local` to enable running the debug command.

Total memory used
32-bit: **2634404kb (2.64gb)**
64-bit: **3020524kb (3.02gb)**

So, ~14.4% more memory efficient. On the surface, yeah, that extra 14.4% would be helpful for resource constrained situations. I haven't done a benchmark, my guess is they'd be pretty comparable, although I'd love to see something on a benchmark.

What I'm a little _more_ concerned about is what the project would need to maintain by issuing a 32-bit container. I'm not sure how much 32-bit Valkey is tested and the long-term outlook for 32-bit stuff is... grim. We're far beyond the days when 32-bit is needed to support any plausible, contemporary hardware. 

64-bit is a far less constrained environment for a memory hungry application like Valkey; I feel like 32-bit containers would be used only by a handful of people and they are a non-trivial thing to maintain.

**Reply by binaryfire** [reply]
All fair points. The use case it would really shine for is IoT devices but yeah, I’m not sure the extra maintenance headaches are worth it.

**Reply by zuiderkwast** [reply]
Interesting! 1MB keys are quite large. For really small keys, the memory difference would be a lot bigger. Do you still have the binary? Could you try DEBUG POPULATE with the default key size?

**Comment by zuiderkwast** [comment]
This is an interesting topic and I'm interested in knowing more about the memory usage for different workloads. For small keys, or hashes with many small elements, I believe it can save a lot of memory. I'd like to improve that documentation section with some up-to-date numbers too.

32-bit is still supported and we have CI jobs for x86 32-bit. It's not a prioritized target though and maintaining containers for various platsforms involves a considerable maintenance burden, so I don't really want to commit to that. We've already noticed a maintenance burden for the containers we already have. For example, we need to compile for different OpenSSL versions for different Ubuntu releases. That said, if someone wants to maintain such containers, I'm not going to be against it.

Recently, we've seen that there are 32-bit VMs that compile to WASM and run in a browser. See this: https://github.com/valkey-io/valkey/issues/1412

**Comment by stockholmux** [comment]
I'd have to rebuild everything @zuiderkwast. It's not _too_ much effort, but I'll need to find some time.
 
But you're right and there are a lot of factors too: this was all strings and other data structures have more pointers which means a bigger difference but that's not too easy to generate.

But I don't think it truly changes the arguments against a 32-bit container. I'm a little curious of the whole idea of 32-bit containers are moribund. And the 32-bit in-browser VM is the first thing I've seen that makes me _not_ want to kill 32-bit Valkey with fire.

(And FWIW, I ran `DEBUG POPULATE` repeatedly until I got to the fill numbers indicated - I don't think that was very clear in my initial write up.)

**Reply by zuiderkwast** [reply]
> I'd have to rebuild everything

Then don't. :)

> other data structures have more pointers which means a bigger difference but that's not too easy to generate.

Many small keys are easy to generate. You did 1MB each, but `DEBUG POPULATE 1000000` creates a million small keys. The ration between pointers and content is higher than for large values...

> But I don't think it truly changes the arguments against a 32-bit container. I'm a little curious of the whole idea of 32-bit containers are moribund. And the 32-bit in-browser VM is the first thing I've seen that makes me not want to kill 32-bit Valkey with fire.

Haha, I think we should distinguish having code that builds on 32-bit (x86 and ARM) and maintaining containers. The former is just about sticking to standard C and not making assumptions about what the size of a pointer is. That's just sound IMO and makes the code easily portable to new hardware and OSes.

I'm not convinced about 32-bit containers either. The places where you'd want a 32-bit build is likely not a place where you want to run containers, like the in-browser-VM or some small embedded devices like Arduino that are still 32-bit ARM it seems (at least some of them).
